packages/serverpod_client/lib/src/client_method_stream_manager.dart

```diff
@@ -0,0 +1,296 @@
+      method: message.method,
+    );
+    var inboundStreamContext = _inboundStreams[inboundStreamKey];
+    if (inboundStreamContext == null) return;
+
+    if (inboundStreamContext.openCompleter.isCompleted) return;
+
+    inboundStreamContext.openCompleter.complete(message.responseType);
+  }
+
+  Future<void> _dispatchSerializableException(
+    MethodStreamSerializableException message,
+  ) async {
+    var inboundStreamKey = _buildStreamKey(
+      connectionId: message.connectionId,
+      endpoint: message.endpoint,
+      method: message.method,
+    );
+
+    var inboundStreamContext = _inboundStreams.remove(inboundStreamKey);
+    if (inboundStreamContext == null) {
+      return;
+    }
+
+    inboundStreamContext.controller.addError(message.exception);
+    await _closeControllers([inboundStreamContext.controller]);
+  }
+
+  Future<void> _listenToWebSocketStream() async {
+    var webSocket = _webSocket;
+    if (webSocket == null) {
+      return;
+    }
+
+    try {
+      await for (String jsonData in webSocket.stream) {
+        if (!_handshakeComplete.isCompleted) {
+          _handshakeComplete.complete();
+        }
+
+        WebSocketMessage message;
+        try {
+          message = WebSocketMessage.fromJsonString(
+            jsonData,
+            _serializationManager,
+          );
+        } on UnknownMessageException catch (_) {
+          _webSocket?.sink.add(BadRequestMessage.buildMessage(jsonData));
+          rethrow;
+        }
+
+        switch (message) {
+          case OpenMethodStreamResponse():
+            _dispatchMethodStreamResponse(message);
+          case OpenMethodStreamCommand():
+            // Ignore
+            break;
+          case CloseMethodStreamCommand():
+            unawaited(_dispatchCloseMethodStreamCommand(message));
+            break;
+          case PingCommand():
+            _webSocket?.sink.add(PongCommand.buildMessage());
+            break;
+          case PongCommand():
+            // Ignore
+            break;
+          case MethodStreamSerializableException():
+            unawaited(_dispatchSerializableException(message));
+            break;
+          case MethodStreamMessage():
+            var success = _dispatchMessage(message);
+            if (success) break;
+
+            _webSocket?.sink.add(
+              CloseMethodStreamCommand.buildMessage(
+                connectionId: message.connectionId,
+                endpoint: message.endpoint,
+                method: message.method,
+                reason: CloseReason.error,
+              ),
+            );
+            break;
+          case BadRequestMessage():
+            throw Exception(
+                'Bad request message: $jsonData, closing connection');
+        }
+      }
+    } catch (e, s) {
+      throw WebSocketListenException(e, s);
+    } finally {
+      closeConnection();
+    }
+  }
+
+  Future<void> _tryConnect() async {
+    // TODO: Ensure this section is not run in parallel.
```

Comment:
Question: Is this handled? If not we should otherwise remove todo.

Reply:
Yes, this is fixed in a later commit.